c-bindings-gen/src/types.rs

```diff
@@ -0,0 +1,1997 @@
+					DeclType::EnumIgnored|DeclType::StructImported if !is_ref && from_ptr =>
+						write!(w, ", _underlying_ref: false }}").unwrap(),
+					DeclType::EnumIgnored|DeclType::StructImported if !is_ref => write!(w, ")), _underlying_ref: false }}").unwrap(),
+					DeclType::Trait(_) if is_ref => {},
+					_ => unimplemented!(),
+				});
+	}
+	pub fn write_to_c_conversion_inline_suffix<W: std::io::Write>(&self, w: &mut W, t: &syn::Type, generics: Option<&GenericTypes>, ptr_for_ref: bool) {
+		self.write_to_c_conversion_inline_suffix_inner(w, t, generics, false, ptr_for_ref, false);
+	}
+
+	fn write_from_c_conversion_prefix_inner<W: std::io::Write>(&self, w: &mut W, t: &syn::Type, generics: Option<&GenericTypes>, is_ref: bool, ptr_for_ref: bool) {
+		self.write_conversion_inline_intern(w, t, generics, is_ref, false, "() /*", true, |_| "&local_",
+				|a, b, _c| self.from_c_conversion_prefix_from_path(a, b),
+				|w, decl_type, _full_path, is_ref| match decl_type {
+					DeclType::StructImported if is_ref && ptr_for_ref => write!(w, "unsafe {{ &*(*").unwrap(),
+					DeclType::StructImported if is_ref => write!(w, "unsafe {{ &mut *").unwrap(),
+					DeclType::StructImported if !is_ref => write!(w, "*unsafe {{ Box::from_raw(").unwrap(),
+					DeclType::MirroredEnum if is_ref => write!(w, "&").unwrap(),
+					DeclType::MirroredEnum => {},
+					DeclType::Trait(_) => {},
+					_ => unimplemented!(),
+				});
+	}
+	pub fn write_from_c_conversion_prefix<W: std::io::Write>(&self, w: &mut W, t: &syn::Type, generics: Option<&GenericTypes>) {
+		self.write_from_c_conversion_prefix_inner(w, t, generics, false, false);
+	}
+	fn write_from_c_conversion_suffix_inner<W: std::io::Write>(&self, w: &mut W, t: &syn::Type, generics: Option<&GenericTypes>, is_ref: bool, ptr_for_ref: bool) {
+		self.write_conversion_inline_intern(w, t, generics, is_ref, false, "*/", false,
+				|has_inner| match has_inner {
+					false => ".iter().collect::<Vec<_>>()[..]",
+					true => "[..]",
+				},
+				|a, b, _c| self.from_c_conversion_suffix_from_path(a, b),
+				|w, decl_type, _full_path, is_ref| match decl_type {
+					DeclType::StructImported if is_ref && ptr_for_ref => write!(w, ").inner }}").unwrap(),
+					DeclType::StructImported if is_ref => write!(w, ".inner }}").unwrap(),
+					DeclType::StructImported if !is_ref => write!(w, ".inner.take_ptr() as *mut _) }}").unwrap(),
+					DeclType::MirroredEnum if is_ref => write!(w, ".to_ln()").unwrap(),
+					DeclType::MirroredEnum => write!(w, ".into_ln()").unwrap(),
+					DeclType::Trait(_) => {},
+					_ => unimplemented!(),
+				});
+	}
+	pub fn write_from_c_conversion_suffix<W: std::io::Write>(&self, w: &mut W, t: &syn::Type, generics: Option<&GenericTypes>) {
+		self.write_from_c_conversion_suffix_inner(w, t, generics, false, false);
+	}
+	// Note that compared to the above conversion functions, the following are generally
+	// significantly undertested:
```

Comment:
I think this means the following two functions? If so, I think it'd be good to be more specific.